### ISSUE TITLE
User specified retries of a workflow step

### ISSUE THREAD

**Issue by jessesuen** [argoproj/argo#498]
One request is the supporting of retries of a workflow step. A clean way we could support this is through the use of a kuberentes Job instead of a Pod. Kubernetes jobs already have built in support for timeouts (activeDeadlineSeconds), retries (backoffLimit). If the user specifies in the workflow spec a timeout or retry, the controller could choose to schedule a job instead of a pod, and let the native job controller handle all the logic of retries, timeouts, etc... 

https://kubernetes.io/docs/concepts/workloads/controllers/jobs-run-to-completion/#job-termination-and-cleanup

**Comment by javierbq** [comment]
@jessesuen I think all steps should be run in Jobs instead of pods in order to ensure execution of steps in the event of nodes coming offline.  This feature would allow using secondary node pools made out of preemptible instances.

It would be great if each pod could inherit timeouts and retry defaults from the flow and that this values could get overwritten at the job level.

**Comment by jessesuen** [comment]
In our previous implementation of workflows in Argo V1, it actually used to be the case that we scheduled containers using jobs instead of pods. What we discovered, is that there is a real and perceptible overhead in using Jobs over Pods, and we would hit scalability issues when scheduling hundreds of jobs and subsequently attempting to monitor them. As soon as we moved away from Jobs to Pods, we were able to achieve much higher scale.

Although it is convenient to "inherit" some of the functionality from jobs and get those features for free, I don't think it is worth paying the scalability cost to schedule everything as jobs.

One thing I recently discovered is that activeDeadlineSeconds is already part of the pod spec (not just job). So I think it will be trivial to incorporate that language to the template spec, carry forward that information to the pod, and get container level timeouts for free. Then the last remaining piece would be to replicate the functionality of retries (backoffLimit) in the same manner that that Job controller does it. But I do not believe this to be that difficult of a task to do. This way we can get the same functionality as Job, without the overhead.

**Comment by javierbq** [comment]
I am having problems with steps that run until completion but fail during the upload of artifacts. The containers are labeled as Completed but the step is marked as failed. Is argo doing any retries on upload of artifacts?

**Comment by jessesuen** [comment]
The pod is supposed to be annotated with an error message with a precise failure reason (e.g. "failed to save artifacts: <some specific reason>". Do you have that message from `argo get <workflowname>`? If not, please share the logs of the wait container:  `argo logs <podname> -c wait` which should have output indicating the reason.

I also want to make sure we are not failing because of conservative memory limits we currently set in the wait sidecar, especially when dealing with large artifacts which can put memory pressure on tar/gz. `kubectl get|describe pod <podname>` would indicate if we were killed because of memory constraints.

Regarding retries, we do not currently do retries on artifact loading/saving.

**Comment by javierbq** [comment]
I am getting random TLS handshake errors in my flows.

```+ argoexec sidecar wait
time="2017-12-18T03:06:36Z" level=info msg="Executor (version: 2.0.0-1069af4) initialized with template:\narchiveLocation:\n  s3:\n    accessKeySecret:\n      key: accessKey\n      name: gcs-credentials\n    bucket: argo-e5119d83289bd3cd10\n    endpoint: storage.googleapis.com\n    key: hm-6zb47/hm-6zb47-734900164\n    secretKeySecret:\n      key: secretKey\n      name: gcs-credentials\ncontainer:\n  args:\n  - /home/sequence.fasta\n  - -a\n  - /tmp/sparksx_cambyses_aln.json\n  command:\n  - run_sparksx.py\n  image: gcr.io/cyrus-playground/sparksx:latest\n  name: \"\"\n  resources: {}\n  volumeMounts:\n  - mountPath: /var/databases\n    name: protein-databases\n    readOnly: true\ninputs:\n  artifacts:\n  - name: fasta\n    path: /home/sequence.fasta\n    s3:\n      accessKeySecret:\n        key: accessKey\n        name: gcs-credentials\n      bucket: argo-e5119d83289bd3cd10\n      endpoint: storage.googleapis.com\n      key: hm-6zb47/hm-6zb47-2698807338/fasta.tgz\n      secretKeySecret:\n        key: secretKey\n        name: gcs-credentials\n  - name: pssm\n    path: /tmp/target.pssm\n    s3:\n      accessKeySecret:\n        key: accessKey\n        name: gcs-credentials\n      bucket: argo-e5119d83289bd3cd10\n      endpoint: storage.googleapis.com\n      key: hm-6zb47/hm-6zb47-1588772319/pssm.tgz\n      secretKeySecret:\n        key: secretKey\n        name: gcs-credentials\nname: sparksx\noutputs:\n  artifacts:\n  - name: sparksx-cambyses-alignment\n    path: /tmp/sparksx_cambyses_aln.json\n"
time="2017-12-18T03:06:36Z" level=info msg="Waiting on main container"
time="2017-12-18T03:06:44Z" level=info msg="[docker wait de395d4c6d3abb22ae0d305b7bbd925edae78f88ee38ba0109475b9e3cdcf877]"
time="2017-12-18T03:12:38Z" level=info msg="Waiting completed"
time="2017-12-18T03:12:38Z" level=info msg="Killing sidecars"
+ argoexec artifacts save
time="2017-12-18T03:17:58Z" level=info msg="Executor (version: 2.0.0-1069af4) initialized with template:\narchiveLocation:\n  s3:\n    accessKeySecret:\n      key: accessKey\n      name: gcs-credentials\n    bucket: argo-e5119d83289bd3cd10\n    endpoint: storage.googleapis.com\n    key: hm-6zb47/hm-6zb47-734900164\n    secretKeySecret:\n      key: secretKey\n      name: gcs-credentials\ncontainer:\n  args:\n  - /home/sequence.fasta\n  - -a\n  - /tmp/sparksx_cambyses_aln.json\n  command:\n  - run_sparksx.py\n  image: gcr.io/cyrus-playground/sparksx:latest\n  name: \"\"\n  resources: {}\n  volumeMounts:\n  - mountPath: /var/databases\n    name: protein-databases\n    readOnly: true\ninputs:\n  artifacts:\n  - name: fasta\n    path: /home/sequence.fasta\n    s3:\n      accessKeySecret:\n        key: accessKey\n        name: gcs-credentials\n      bucket: *********\n      endpoint: storage.googleapis.com\n      key: hm-6zb47/hm-6zb47-2698807338/fasta.tgz\n      secretKeySecret:\n        key: secretKey\n        name: gcs-credentials\n  - name: pssm\n    path: /tmp/target.pssm\n    s3:\n      accessKeySecret:\n        key: accessKey\n        name: gcs-credentials\n      bucket: argo-e5119d83289bd3cd10\n      endpoint: storage.googleapis.com\n      key: hm-6zb47/hm-6zb47-1588772319/pssm.tgz\n      secretKeySecret:\n        key: secretKey\n        name: gcs-credentials\nname: sparksx\noutputs:\n  artifacts:\n  - name: sparksx-cambyses-alignment\n    path: /tmp/sparksx_cambyses_aln.json\n"
time="2017-12-18T03:17:58Z" level=info msg="Saving output artifacts"
time="2017-12-18T03:17:58Z" level=info msg="Main container identified as: de395d4c6d3abb22ae0d305b7bbd925edae78f88ee38ba0109475b9e3cdcf877"
time="2017-12-18T03:17:58Z" level=info msg="Saving artifact: sparksx-cambyses-alignment"
time="2017-12-18T03:17:58Z" level=info msg="Archiving de395d4c6d3abb22ae0d305b7bbd925edae78f88ee38ba0109475b9e3cdcf877:/tmp/sparksx_cambyses_aln.json to /argo/outputs/artifacts/sparksx-cambyses-alignment.tgz"
time="2017-12-18T03:17:58Z" level=info msg="[sh -c docker cp -a de395d4c6d3abb22ae0d305b7bbd925edae78f88ee38ba0109475b9e3cdcf877:/tmp/sparksx_cambyses_aln.json - | gzip > /argo/outputs/artifacts/sparksx-cambyses-alignment.tgz]"
time="2017-12-18T03:18:02Z" level=info msg="Archiving completed"
time="2017-12-18T03:18:03Z" level=info msg="Saving from /argo/outputs/artifacts/sparksx-cambyses-alignment.tgz to s3 (endpoint: storage.googleapis.com, bucket: argo-e5119d83289bd3cd10, key: hm-6zb47/hm-6zb47-734900164/sparksx-cambyses-alignment.tgz)"
time="2017-12-18T03:19:00Z" level=fatal msg="Error saving output artifacts, [ERR_INTERNAL] Get https://storage.googleapis.com/argo-e5119d83289bd3cd10/?location=: net/http: TLS handshake timeout\ngithub.com/argoproj/argo/errors.Wrap\n\t/root/go/src/github.com/argoproj/argo/errors/errors.go:86\ngithub.com/argoproj/argo/errors.InternalWrapError\n\t/root/go/src/github.com/argoproj/argo/errors/errors.go:69\ngithub.com/argoproj/argo/workflow/artifacts/s3.(*S3ArtifactDriver).Save\n\t/root/go/src/github.com/argoproj/argo/workflow/artifacts/s3/s3.go:53\ngithub.com/argoproj/argo/workflow/executor.(*WorkflowExecutor).SaveArtifacts\n\t/root/go/src/github.com/argoproj/argo/workflow/executor/executor.go:174\ngithub.com/argoproj/argo/cmd/argoexec/commands.saveArtifacts\n\t/root/go/src/github.com/argoproj/argo/cmd/argoexec/commands/artifacts.go:143\ngithub.com/argoproj/argo/vendor/github.com/spf13/cobra.(*Command).execute\n\t/root/go/src/github.com/argoproj/argo/vendor/github.com/spf13/cobra/command.go:702\ngithub.com/argoproj/argo/vendor/github.com/spf13/cobra.(*Command).ExecuteC\n\t/root/go/src/github.com/argoproj/argo/vendor/github.com/spf13/cobra/command.go:783\ngithub.com/argoproj/argo/vendor/github.com/spf13/cobra.(*Command).Execute\n\t/root/go/src/github.com/argoproj/argo/vendor/github.com/spf13/cobra/command.go:736\nmain.main\n\t/root/go/src/github.com/argoproj/argo/cmd/argoexec/main.go:11\nruntime.main\n\t/usr/local/go/src/runtime/proc.go:185\nruntime.goexit\n\t/usr/local/go/src/runtime/asm_amd64.s:2337"```

**Comment by jessesuen** [comment]
This error is coming from the s3 client library we are using (minio-go). I split out this issue into a new issue #612 because there is a difference in retries for some core workflow execution functionality (saving artifacts) vs. a this issue for a feature to allow user specified retries of a step. The former we would want to do irrespective of the latter.

**Comment by jessesuen** [comment]
I also split out the ability to timeout a container step into a separate issue #613 which will be implemented earlier than the user specified retries. Code is in pull #608.

**Comment by jessesuen** [comment]
Retry is now in master with commit 290f679
